src/mse.cu

```diff
@@ -0,0 +1,602 @@
+            if (i + blockSize < n) {
+                mySum += mse_loss(output[(i + blockSize) * incx], labels[(i + blockSize) * incx]);
+            }
+
+            i += gridSize;
+        }
+    }
+
+    shared_data[tid] = mySum / n;
+
+    __syncthreads();
+
+    sum_reduce_impl<T, blockSize>(r_output, shared_data);
+}
+
+template <class T, size_t blockSize, bool Reduce>
+__global__ void mse_loss_kernel1(size_t n, const T* output, const T* labels, T* r_output) {
+    extern __shared__ volatile unsigned char shared_data_raw[];
+
+    volatile T* shared_data = reinterpret_cast<volatile T*>(shared_data_raw);
+
+    size_t tid      = threadIdx.x;
+    size_t i        = blockIdx.x * (2 * blockDim.x) + threadIdx.x;
+    size_t gridSize = blockSize * 2 * gridDim.x;
+
+    // Perform first level of reduction,
+    // reading from global memory and writing to shared memory
+
+    T mySum = 0;
+
+    if (Reduce) {
+        // In case of reductions, this a simple sum (labels are ignored)
+
+        while (i < n) {
+            mySum += output[i];
+
+            if (i + blockSize < n) {
+                mySum += output[i + blockSize];
+            }
+
+            i += gridSize;
+        }
+    } else {
+        // In the basic case, perform reduction and MSE loss
+
+        while (i < n) {
+            mySum += mse_loss(output[i], labels[i]);
+
+            if (i + blockSize < n) {
+                mySum += mse_loss(output[i + blockSize], labels[i + blockSize]);
+            }
+
+            i += gridSize;
+        }
+    }
+
+    shared_data[tid] = mySum / n;
+
+    __syncthreads();
+
+    sum_reduce_impl<T, blockSize>(r_output, shared_data);
+}
+
+template <class T, size_t blockSize, bool Reduce>
+__global__ void mse_error_kernel(size_t n, const T* output, size_t incx, const T* labels, size_t incy, T* r_output) {
+    extern __shared__ volatile unsigned char shared_data_raw[];
+
+    volatile T* shared_data = reinterpret_cast<volatile T*>(shared_data_raw);
+
+    size_t tid      = threadIdx.x;
+    size_t i        = blockIdx.x * (2 * blockDim.x) + threadIdx.x;
+    size_t gridSize = blockSize * 2 * gridDim.x;
+
+    // Perform first level of reduction,
+    // reading from global memory and writing to shared memory
+
+    T mySum = 0;
+
+    if (Reduce) {
+        // In case of reductions, this a simple sum (labels are ignored)
+
+        while (i < n) {
+            mySum += output[i * incx];
+
+            if (i + blockSize < n) {
+                mySum += output[(i + blockSize) * incx];
+            }
+
+            i += gridSize;
+        }
+    } else {
+        // In the basic case, perform reduction and MSE error
+
+        while (i < n) {
+            mySum += mse_error(output[i * incx], labels[i * incy]);
+
+            if (i + blockSize < n) {
+                mySum += mse_error(output[(i + blockSize) * incx], labels[(i + blockSize) * incx]);
+            }
+
+            i += gridSize;
+        }
+    }
+
+    shared_data[tid] = mySum / n;
```

Comment:
Why did you divide by `n` here? By doing this, the implementation will not be equivalent to the one from DLL. The scaling is done by using the `alpha `and `beta `factors.